core/consensus/babe/primitives/src/lib.rs

```diff
@@ -49,15 +56,25 @@ pub type SlotNumber = u64;
 /// The weight of an authority.
 pub type Weight = u64;
 
+/// BABE epoch information
+#[derive(Decode, Encode, Default, PartialEq, Eq, Clone)]
+#[cfg_attr(any(feature = "std", test), derive(Debug))]
+pub struct Epoch {
+	/// The authorities and their weights
+	pub authorities: Vec<(AuthorityId, Weight)>,
+	/// Randomness for this epoch
+	pub randomness: [u8; VRF_OUTPUT_LENGTH],
+}
+
 /// An consensus log item for BABE.
-#[derive(Decode, Encode)]
+#[derive(Decode, Encode, Clone, PartialEq, Eq)]
 pub enum ConsensusLog {
 	/// The epoch has changed. This provides information about the
 	/// epoch _after_ next: what slot number it will start at, who are the authorities (and their weights)
 	/// and the next epoch randomness. The information for the _next_ epoch should already
 	/// be available.
 	#[codec(index = "1")]
-	NextEpochData(SlotNumber, Vec<(AuthorityId, Weight)>, [u8; VRF_OUTPUT_LENGTH]),
+	NextEpochData(Epoch),
```

Comment:
The removal of `SlotNumber` here means you can't change the epoch length without a hard fork. And yet, the SRML code doesn't account for a constant `slots_per_epoch`. This is probably a bad idea. I would favor a dynamic `SlotNumber` in the `NextEpochData` digest which is the earliest slot number we can start the epoch after the one started by the digest.

Reply:
@rphmeier is making `SlotNumber` part of `Epoch` sufficient?

Reply:
as long as it informs the client, yes.